File: code/modM_DMSepConstr.R
# Constrained 2n-2 BYM2 model for M_DM (DHS+MICS fusion) data
# Uses n-1 free params for w and n-1 free params for u,
# deriving the nth element as -sum(free) to enforce sum-to-zero.
# Explicit alpha intercept parameter.

fitMDMConstr = function(datDHS=ed, datMICS=edMICS, inputsMDM=NULL, 
                        intPtsMICS=NULL, intPtsDHS=NULL, 
                        KMICS=100,
                        KDHSurb = 11,
                        JInnerUrban = 3,
                        KDHSrur = 16,
                        JInnerRural = 3,
                        JOuterRural = 1, admMICS=admFinal, adm2DHS=adm2Full, 
                        alpha_pri = c(0, 100^2), 
                        beta_pri = c(0, sqrt(1000)), 
                        pc.bym2Phi=list(u=0.5, alpha=1/3), 
                        pc.bym2Prec=list(u=1, alpha=0.5), 
                        pc.expPrec=list(u=1, alpha=0.5), 
                        maxit=1000, adm2AsCovariate=FALSE, 
                        doMCMC=FALSE) {
  
  # make sure Stratum variable exists in MICS data
  if(!("Stratum" %in% names(datMICS))) {
    datMICS$Stratum = adm2ToStratumMICS(datMICS$subarea)
  }
  
  # make sure all the names in our datasets we expect to be there are
  nameTab = rbind(c("N", "ns"), 
                  c("N", "n"), 
                  c("Z", "ys"), 
                  c("Z", "y"))
  for(i in 1:nrow(nameTab)) {
    theseNames = nameTab[i,]
    fromN = theseNames[1]
    toN = theseNames[2]
    if(!(toN %in% names(datMICS))) {
      datMICS[[toN]] = datMICS[[fromN]]
    }
    if(!(toN %in% names(datDHS))) {
      datDHS[[toN]] = datDHS[[fromN]]
    }
  }
  
  # first generate all necessary inputs if need be
  if(is.null(inputsMDM)) {
    print("Making M_DM inputs...")
    
    inputsMDM = makeInputsMDM(datDHS, datMICS, 
                              intPtsMICS=intPtsMICS, intPtsDHS=intPtsDHS, 
                              KMICS=KMICS,
                              KDHSurb = KDHSurb,
                              JInnerUrban = JInnerUrban,
                              KDHSrur = KDHSrur,
                              JInnerRural = JInnerRural,
                              JOuterRural = JOuterRural, 
                              admMICS=admMICS, adm2DHS=adm2DHS, 
                              adm2AsCovariate=adm2AsCovariate)
    
  }
  
  thisEnv = environment()
  list2env(inputsMDM, envir=thisEnv)
  
  datMICS = sortByCol(datMICS, "Stratum", admMICS$NAME_FINAL)
  
  # set priors ----
  
  out = load("savedOutput/global/admFinalMat.RData")
  bym2ArgsTMB = prepareBYM2argumentsForTMB(admFinalMat, u=pc.bym2Phi$u, alpha=pc.bym2Phi$alpha, 
                                           constr=TRUE, scale.model=TRUE, matrixType="TsparseMatrix")
  lambdaTau = getLambdaPCprec(u=pc.bym2Prec$u, alpha=pc.bym2Prec$alpha)
  lambdaTauEps = getLambdaPCprec(u=pc.expPrec$u, alpha=pc.expPrec$alpha)
  
  nAreas = ncol(bym2ArgsTMB$Q)
  
  # Specify inputs for TMB ----
  
  ## specify random effects
  rand_effs <- c('alpha', 'beta', 'w_bym2Free', 'u_bym2Free', 
                 'nuggetUrbMICS', 'nuggetRurMICS', 'nuggetUrbDHS', 'nuggetRurDHS')
  
  # collect input data
  data_full = list(
    y_iUrbanMICS=ysUrbMICS,
    y_iRuralMICS=ysRurMICS,
    n_iUrbanMICS=nsUrbMICS,
    n_iRuralMICS=nsRurMICS,
    areaidxlocUrbanMICS=areaidxlocUrbanMICS,
    areaidxlocRuralMICS=areaidxlocRuralMICS,
    X_betaUrbanMICS=intPtsMICS$XUrb,
    X_betaRuralMICS=intPtsMICS$XRur,
    wUrbanMICS=intPtsMICS$wUrban,
    wRuralMICS=intPtsMICS$wRural,
    
    y_iUrbanDHS=ysUrbDHS,
    y_iRuralDHS=ysRurDHS,
    n_iUrbanDHS=nsUrbDHS,
    n_iRuralDHS=nsRurDHS,
    areaidxlocUrbanDHS=areaidxlocUrbanDHS,
    areaidxlocRuralDHS=areaidxlocRuralDHS,
    X_betaUrbanDHS=intPtsDHS$covsUrb,
    X_betaRuralDHS=intPtsDHS$covsRur,
    wUrbanDHS=intPtsDHS$wUrban,
    wRuralDHS=intPtsDHS$wRural,
    
    Q_bym2=bym2ArgsTMB$Q,
    alpha_pri=alpha_pri,
    beta_pri=beta_pri,
    tr=bym2ArgsTMB$tr,
    gammaTildesm1=bym2ArgsTMB$gammaTildesm1,
    lambdaPhi=bym2ArgsTMB$lambda,
    lambdaTau=lambdaTau,
    lambdaTauEps=lambdaTauEps, 
    options=0
  )
  
  
  anyna = function(x) {any(is.na(x))}
  myDim = function(x) {
    dims = dim(x)
    if(is.null(dims)) {
      length(x)
    }
    else {
      dims
    }
  }
  
  # initial parameters
  initUrbP = sum(c(data_full$y_iUrbanMICS, data_full$y_iUrbanDHS))/sum(c(data_full$n_iUrbanMICS, data_full$n_iUrbanDHS))
  initRurP = sum(c(data_full$y_iRuralMICS, data_full$y_iRuralDHS))/sum(c(data_full$n_iRuralMICS, data_full$n_iRuralDHS))
  initAlpha = logit(initRurP)
  initBeta1 = logit(initUrbP) - initAlpha
  
  tmb_params <- list(log_tau = 0,
                     logit_phi = 0,
                     log_tauEps = 0,
                     alpha = initAlpha,
                     beta = c(initBeta1, rep(0, ncol(intPtsDHS$covsUrb)-1)), 
                     w_bym2Free = rep(0, nAreas - 1),
                     u_bym2Free = rep(0, nAreas - 1),
                     nuggetUrbMICS = rep(0, length(data_full$y_iUrbanMICS)), 
                     nuggetRurMICS = rep(0, length(data_full$y_iRuralMICS)),
                     nuggetUrbDHS = rep(0, length(data_full$y_iUrbanDHS)), 
                     nuggetRurDHS = rep(0, length(data_full$y_iRuralDHS))
  )
  
  
  # make TMB fun and grad ----
  
  # first make sure no TMB dynlib is loaded
  unloadDynlibs()
  
  # compile dynlib if need be
  if(!any(file.exists(paste0("code/modM_DMSepConstr", c(".o", ".so", ".dll"))))) {
    print("compiling code/modM_DMSepConstr.cpp...")
    compile( "code/modM_DMSepConstr.cpp", 
             framework="TMBad", safebounds=FALSE)
  }
  
  # load dynlib, make tmb obj
  dyn.load( dynlib("code/modM_DMSepConstr"))
  TMB::config(tmbad.sparse_hessian_compress = 1)
  obj <- MakeADFun(data=data_full,
                   parameters=tmb_params,
                   random=rand_effs,
                   hessian=TRUE,
                   DLL='modM_DMSepConstr')
  
  lower = rep(-10, length(obj[['par']]))
  upper = rep( 10, length(obj[['par']]))
  
  # make wrapper functions that print out parameters and function values
  funWrapper = function(par, badParVal=1e10) {
    if(any(par < lower) || any(par > upper)) {
      return(badParVal)
    }
    print(par)
    objVal = testObj[['fn']](par)
    parNames = names(par)
    parVals = par
    parStrs = sapply(1:length(par), function(ind) {paste(parNames[ind], ": ", parVals[ind], sep="")})
    parStr = paste(parStrs, collapse=", ")
    print(paste0("objective: ", objVal, " for parameters, ", parStr))
    objVal
  }
  
  grWrapper = function(par, badParVal=1e10) {
    if(any(par < lower) || any(par > upper)) {
      return(rep(badParVal, length(par)))
    }
    print(par)
    grVal = testObj[['gr']](par)
    parNames = names(par)
    parVals = par
    parStrs = sapply(1:length(par), function(ind) {paste(parNames[ind], ": ", parVals[ind], sep="")})
    parStr = paste(parStrs, collapse=", ")
    grStr = paste(grVal, collapse=", ")
    print(paste0("gradient: ", grStr, " for parameters, ", parStr))
    grVal
  }
  
  # * Run TMB ----
  
  if(!doMCMC) {
    {
      tolSeq = 1e-06
      testObj = obj
      optPar = testObj$par
      startTime = proc.time()[3]
      for(thisTol in tolSeq) {
        testObj = obj
        testObj$env$inner.control = list(maxit=maxit, tol10=thisTol)
        testObj$env$tracepar = TRUE
        print(paste0("optimizing for tol = ", thisTol, "."))
        opt1 <- optim(par=optPar, fn=funWrapper, gr=grWrapper,
                      method = c("BFGS"), hessian = FALSE, control=list(reltol=thisTol))
        optPar = opt1$par
        if(!is.null(opt1$message)) {
          print(paste0("error for tol = ", thisTol, ". Message:"))
          print(opt1$message)
          next
        }
        else {
          print(paste0("completed optimization for tol = ", thisTol, ""))
          
          ## Get standard errors
          print("getting standard errors...")
          sdTime = system.time(
            SD0 <- TMB::sdreport(testObj, getJointPrecision=TRUE,
                                 bias.correct = TRUE,
                                 bias.correct.control = list(sd = TRUE))
          )[3]
          print(sdTime/60)
          
          if(SD0$pdHess) {
            print("Optimization and PD hess calculation done!")
            break
          }
          else {
            print("Hessan not PD. Rerunning optimization with stricter tol...")
          }
        }
      }
      endTime = proc.time()[3]
      sdTime/60
      totalTime = endTime - startTime
      print(paste0("optimization took ", totalTime/60, " minutes"))
    }
  } else {
    {
      testObj = obj
      optPar = testObj$par
      startTime = proc.time()[3]
      
      fit <- tmbstan(obj=obj, silent=FALSE, laplace=FALSE)
      
      endTime = proc.time()[3]
      totalTime = endTime - startTime
      print(paste0("MCMC took ", totalTime/60, " minutes"))
    }
  }
  
  
  list(TMBobj=obj, TMBsd=SD0, totalTime=totalTime, sdTime=sdTime)
}

File: code/modM_DMSepConstr.cpp
// Constrained 2n-2 BYM2 model for M_DM (DHS+MICS fusion) data
// Uses n-1 free params for w and n-1 free params for u,
// deriving the nth element as -sum(free) to enforce sum-to-zero constraints.
// Explicit alpha intercept parameter.
//
// include libraries
#include <TMB.hpp>
#include <Eigen/Sparse>
#include <vector>
using namespace density;
using Eigen::SparseMatrix;

// helper function for detecting NAs in the data supplied from R
template<class Type>
bool isNA(Type x){
  return R_IsNA(asDouble(x));
}

// PC prior on BYM2 tau (precision) parameter (on log scale)
template<class Type>
Type dPCPriTau(Type logTau, Type lambda)
{
  Type tau = exp(logTau);
  
  Type ldensity = log(lambda) - log(2.0) -Type(1.5)*logTau - lambda/sqrt(tau);
  
  // add in log Jacobian
  Type ljacobian = logTau;
  
  // get final log density
  Type lPriorTau = ldensity + ljacobian;
  
  return lPriorTau;
}

///////////////////////////
// the main function     //
// to calculate the jnll //
///////////////////////////
template<class Type>
Type objective_function<Type>::operator() ()
{
  
  // ~~~~~~~~~------------------------------------------------------~~
  // FIRST, we define params/values/data that will be passed in from R
  // ~~~~~~~~~~~------------------------------------------------------
  
  // MICS data
  DATA_VECTOR( y_iUrbanMICS );
  DATA_VECTOR( y_iRuralMICS );
  DATA_VECTOR( n_iUrbanMICS );
  DATA_VECTOR( n_iRuralMICS );
  DATA_IVECTOR(areaidxlocUrbanMICS);
  DATA_IVECTOR(areaidxlocRuralMICS);
  DATA_MATRIX( X_betaUrbanMICS );
  DATA_MATRIX( X_betaRuralMICS );
  DATA_ARRAY( wUrbanMICS );
  DATA_ARRAY( wRuralMICS );
  
  // DHS data
  DATA_VECTOR( y_iUrbanDHS );
  DATA_VECTOR( y_iRuralDHS );
  DATA_VECTOR( n_iUrbanDHS );
  DATA_VECTOR( n_iRuralDHS );
  DATA_IVECTOR(areaidxlocUrbanDHS);
  DATA_IVECTOR(areaidxlocRuralDHS);
  DATA_MATRIX( X_betaUrbanDHS );
  DATA_MATRIX( X_betaRuralDHS );
  DATA_ARRAY( wUrbanDHS );
  DATA_ARRAY( wRuralDHS );
  
  DATA_SPARSE_MATRIX( Q_bym2 ); // precision matrix for Besag component
  
  // prior parameters
  DATA_VECTOR( alpha_pri );
  DATA_VECTOR( beta_pri );
  
  // BYM2 and prior precomputed values
  DATA_SCALAR( tr );
  DATA_VECTOR( gammaTildesm1 );
  
  DATA_SCALAR( lambdaPhi );
  DATA_SCALAR( lambdaTau );
  DATA_SCALAR( lambdaTauEps );
  
  // Options
  DATA_SCALAR( options );
  
  // Fixed effects
  PARAMETER( log_tau );
  PARAMETER( logit_phi );
  PARAMETER( log_tauEps );
  
  // Intercept (explicit)
  PARAMETER( alpha );
  PARAMETER_VECTOR( beta );
  
  // BYM2 effects: n-1 free parameters each
  PARAMETER_VECTOR( w_bym2Free ); // length n-1
  PARAMETER_VECTOR( u_bym2Free ); // length n-1
  PARAMETER_VECTOR( nuggetUrbMICS );
  PARAMETER_VECTOR( nuggetRurMICS );
  PARAMETER_VECTOR( nuggetUrbDHS );
  PARAMETER_VECTOR( nuggetRurDHS );
  
  // ~~~~~~~~~------------------------------------------------~~
  // SECOND, we define all other objects that we need internally
  // ~~~~~~~~~------------------------------------------------~~
  
  int num_iUrbanMICS = y_iUrbanMICS.size();
  int num_iRuralMICS = y_iRuralMICS.size();
  int num_iUrbanDHS = y_iUrbanDHS.size();
  int num_iRuralDHS = y_iRuralDHS.size();
  int nFree = w_bym2Free.size();
  int nAreas = nFree + 1;
  int n_integrationPointsUrbanMICS = wUrbanMICS.cols();
  int n_integrationPointsRuralMICS = wRuralMICS.cols();
  int n_integrationPointsUrbanDHS = wUrbanDHS.cols();
  int n_integrationPointsRuralDHS = wRuralDHS.cols();
  
  // Transform parameters
  Type tau = exp(log_tau);
  Type phi = Type(1.0)/(Type(1.0) + exp(-logit_phi));
  Type sigmaEps = exp(Type(-0.5) * log_tauEps);
  
  // Reconstruct full n-vectors with sum-to-zero constraint
  vector<Type> w_bym2(nAreas);
  vector<Type> u_bym2(nAreas);
  Type wSum = 0.0;
  Type uSum = 0.0;
  for(int i = 0; i < nFree; i++) {
    w_bym2(i) = w_bym2Free(i);
    u_bym2(i) = u_bym2Free(i);
    wSum += w_bym2Free(i);
    uSum += u_bym2Free(i);
  }
  w_bym2(nFree) = -wSum;
  u_bym2(nFree) = -uSum;
  
  Type jnll = 0.0;
  
  // ~~~~~~~~~------------------------------------------------~~-
  // THIRD, we calculate the contribution to the likelihood from:
  // 1) BYM2 GMRF density
  // 2) priors
  // 3) likelihood
  // ~~~~~~~~~------------------------------------------------~~-
  
  /////////
  // (1) //
  /////////
  
  // Quadratic form using full n-vectors
  
  // tau/(1-phi) w^T w
  Type quadW = 0.0;
  for(int i = 0; i < nAreas; i++) {
    quadW += w_bym2(i) * w_bym2(i);
  }
  quadW = quadW * tau/(Type(1.0)-phi);
  
  // u^T Q u + phi/(1-phi) u^T u
  matrix<Type> transformedU = Q_bym2 * u_bym2.matrix();
  Type quadU = 0.0;
  Type fac = phi/(Type(1.0)-phi);
  for(int i = 0; i < nAreas; i++) {
    quadU += transformedU(i) * u_bym2(i) + fac * u_bym2(i) * u_bym2(i);
  }
  
  // -2 sqrt(phi*tau)/(1-phi) u^T w
  Type quadWU = 0.0;
  for(int i = 0; i < nAreas; i++) {
    quadWU += u_bym2(i) * w_bym2(i);
  }
  quadWU = quadWU * (-Type(2.0) * sqrt(phi * tau)/(Type(1.0)-phi));
  
  Type quadSum = quadW + quadU + quadWU;
  
  // Log determinant for the constrained model: (n-1)*log((1-phi)/tau)
  Type logDet = Type(0.0);
  for(int i = 0; i < gammaTildesm1.size(); i++) {
    logDet += log(Type(1.0) + phi*gammaTildesm1(i));
  }
  Type logDetTau = Type(nFree) * log((Type(1.0) - phi)/tau);
  
  Type bym2LogLik = Type(-0.5) * logDetTau;
  bym2LogLik += Type(-0.5) * quadSum;
  jnll -= bym2LogLik;
  
  // iid nugget model
  Type nuggetLogLik = 0.0;
  for(int i = 0; i< nuggetUrbMICS.size(); i++) {
    nuggetLogLik += dnorm(nuggetUrbMICS(i), Type(0), sigmaEps, true);
  }
  for(int i = 0; i< nuggetRurMICS.size(); i++) {
    nuggetLogLik += dnorm(nuggetRurMICS(i), Type(0), sigmaEps, true);
  }
  for(int i = 0; i< nuggetUrbDHS.size(); i++) {
    nuggetLogLik += dnorm(nuggetUrbDHS(i), Type(0), sigmaEps, true);
  }
  for(int i = 0; i< nuggetRurDHS.size(); i++) {
    nuggetLogLik += dnorm(nuggetRurDHS(i), Type(0), sigmaEps, true);
  }
  jnll -= nuggetLogLik;
  
  /////////
  // (2) //
  /////////
  
  // PC prior for tau
  Type logPriTauEps = dPCPriTau(log_tauEps, lambdaTauEps);
  
  Type firstPt = log(lambdaTau) - log(2.0);
  Type secondPt = -Type(1.5)*log_tau;
  Type thirdPt = - lambdaTau/sqrt(tau);
  Type ldensityTau = firstPt + secondPt + thirdPt;
  Type ljacobianTau = log_tau;
  Type logPriTau = ldensityTau + ljacobianTau;
  
  // PC prior for phi
  Type n = gammaTildesm1.size();
  Type KLD = Type(0.5) * (phi * tr - phi * n - logDet);
  Type d = sqrt(Type(2.0) * KLD);
  Type lexpDensity = log(lambdaPhi) - lambdaPhi * d;
  
  Type sumVal = 0.0;
  for(int i = 0; i < n; i++) {
    sumVal += gammaTildesm1(i)/(1 + gammaTildesm1(i) * phi);
  }
  Type ljacobian = - log(d) - log(2.0) + log(tr - n - sumVal);
  ljacobian = ljacobian - logit_phi - Type(2.0) * log(Type(1.0) + exp(-logit_phi));
  Type logPriPhi = lexpDensity + ljacobian;
  
  jnll -= logPriTau;
  jnll -= logPriTauEps;
  jnll -= logPriPhi;
  
  // prior for intercept
  jnll -= dnorm(alpha, alpha_pri(0), alpha_pri(1), true);
  
  // prior for covariates
  for(int i = 0; i < beta.size(); i++) {
    jnll -= dnorm(beta(i), beta_pri(0), beta_pri(1), true);
  }
  
  /////////
  // (3) //
  /////////
  
  // Fixed effects at data points
  vector<Type> fe_iUrbanMICS = X_betaUrbanMICS * beta;
  vector<Type> fe_iRuralMICS = X_betaRuralMICS * beta;
  vector<Type> fe_iUrbanDHS = X_betaUrbanDHS * beta;
  vector<Type> fe_iRuralDHS = X_betaRuralDHS * beta;
  
  vector<Type> latentFieldUrbMICS(num_iUrbanMICS * n_integrationPointsUrbanMICS);
  vector<Type> latentFieldRurMICS(num_iRuralMICS * n_integrationPointsRuralMICS);
  vector<Type> latentFieldUrbDHS(num_iUrbanDHS * n_integrationPointsUrbanDHS);
  vector<Type> latentFieldRurDHS(num_iRuralDHS * n_integrationPointsRuralDHS);
  int thisIndex;
  Type thisWeight;
  Type thislik;
  array<Type> liksUrbMICS(num_iUrbanMICS, n_integrationPointsUrbanMICS);
  array<Type> liksRurMICS(num_iRuralMICS, n_integrationPointsRuralMICS);
  array<Type> liksUrbDHS(num_iUrbanDHS, n_integrationPointsUrbanDHS);
  array<Type> liksRurDHS(num_iRuralDHS, n_integrationPointsRuralDHS);
  
  // MICS Urban
  for (int obsI = 0; obsI < num_iUrbanMICS; obsI++) {
    thislik = 0.0;
    for (int intI = 0; intI < n_integrationPointsUrbanMICS; intI++) {
      thisIndex = num_iUrbanMICS * intI + obsI;
      latentFieldUrbMICS(thisIndex) = alpha + fe_iUrbanMICS(thisIndex) + w_bym2(areaidxlocUrbanMICS(obsI)) + nuggetUrbMICS(obsI);
      thisWeight = wUrbanMICS(obsI,intI);
      if(thisWeight > 0.0) {
        liksUrbMICS(obsI, intI) = dbinom_robust( y_iUrbanMICS(obsI), n_iUrbanMICS(obsI), latentFieldUrbMICS(thisIndex), false);
        thislik += thisWeight*liksUrbMICS(obsI, intI);
      } else {
        liksUrbMICS(obsI, intI) = thisWeight;
      }
    }
    if(thislik > 0.0) {
      jnll -= log(thislik);
    }
  }
  
  // MICS Rural
  for (int obsI = 0; obsI < num_iRuralMICS; obsI++) {
    thislik = 0.0;
    for (int intI = 0; intI < n_integrationPointsRuralMICS; intI++) {
      thisIndex = num_iRuralMICS * intI + obsI;
      latentFieldRurMICS(thisIndex) = alpha + fe_iRuralMICS(thisIndex) + w_bym2(areaidxlocRuralMICS(obsI)) + nuggetRurMICS(obsI);
      thisWeight = wRuralMICS(obsI,intI);
      if(thisWeight > 0.0) {
        liksRurMICS(obsI, intI) = dbinom_robust( y_iRuralMICS(obsI), n_iRuralMICS(obsI), latentFieldRurMICS(thisIndex), false);
        thislik += thisWeight*liksRurMICS(obsI, intI);
      } else {
        liksRurMICS(obsI, intI) = thisWeight;
      }
    }
    if(thislik > 0.0) {
      jnll -= log(thislik);
    }
  }
  
  // DHS Urban
  for (int obsI = 0; obsI < num_iUrbanDHS; obsI++) {
    thislik = 0.0;
    for (int intI = 0; intI < n_integrationPointsUrbanDHS; intI++) {
      thisIndex = num_iUrbanDHS * intI + obsI;
      latentFieldUrbDHS(thisIndex) = alpha + fe_iUrbanDHS(thisIndex) + w_bym2(areaidxlocUrbanDHS(obsI)) + nuggetUrbDHS(obsI);
      thisWeight = wUrbanDHS(obsI,intI);
      if(thisWeight > 0.0) {
        liksUrbDHS(obsI, intI) = dbinom_robust( y_iUrbanDHS(obsI), n_iUrbanDHS(obsI), latentFieldUrbDHS(thisIndex), false);
        thislik += thisWeight*liksUrbDHS(obsI, intI);
      } else {
        liksUrbDHS(obsI, intI) = thisWeight;
      }
    }
    if(thislik > 0.0) {
      jnll -= log(thislik);
    }
  }
  
  // DHS Rural
  for (int obsI = 0; obsI < num_iRuralDHS; obsI++) {
    thislik = 0.0;
    for (int intI = 0; intI < n_integrationPointsRuralDHS; intI++) {
      thisIndex = num_iRuralDHS * intI + obsI;
      latentFieldRurDHS(thisIndex) = alpha + fe_iRuralDHS(thisIndex) + w_bym2(areaidxlocRuralDHS(obsI)) + nuggetRurDHS(obsI);
      thisWeight = wRuralDHS(obsI,intI);
      if(thisWeight > 0.0) {
        liksRurDHS(obsI, intI) = dbinom_robust( y_iRuralDHS(obsI), n_iRuralDHS(obsI), latentFieldRurDHS(thisIndex), false);
        thislik += thisWeight*liksRurDHS(obsI, intI);
      } else {
        liksRurDHS(obsI, intI) = thisWeight;
      }
    }
    if(thislik > 0.0) {
      jnll -= log(thislik);
    }
  }
  
  // ~~~~~~~~~~~
  // ADREPORT
  // ~~~~~~~~~~~
  if(options==1){
    ADREPORT(log_tau);
    ADREPORT(logit_phi);
  }
  
  REPORT(w_bym2);
  REPORT(u_bym2);
  REPORT(log_tau);
  REPORT(quadU);
  REPORT(quadW);
  REPORT(quadWU);
  REPORT(quadSum);
  REPORT(logDetTau);
  REPORT(bym2LogLik);
  REPORT(nuggetLogLik);
  REPORT(logPriTau);
  REPORT(ldensityTau);
  REPORT(ljacobianTau);
  REPORT(firstPt);
  REPORT(secondPt);
  REPORT(thirdPt);
  REPORT(logPriTauEps);
  REPORT(KLD);
  REPORT(lexpDensity);
  REPORT(ljacobian);
  REPORT(d);
  REPORT(logPriPhi);
  REPORT(liksUrbMICS);
  REPORT(liksRurMICS);
  REPORT(liksUrbDHS);
  REPORT(liksRurDHS);
  REPORT(latentFieldUrbMICS);
  REPORT(latentFieldRurMICS);
  REPORT(latentFieldUrbDHS);
  REPORT(latentFieldRurDHS);
  REPORT(fe_iUrbanMICS);
  REPORT(fe_iRuralMICS);
  REPORT(fe_iUrbanDHS);
  REPORT(fe_iRuralDHS);
  
  
  return jnll;
  
}
